### ISSUE TITLE
[Windows] - Circa 10px strip above tabs

### ISSUE THREAD

**Issue by Anavnaj** [minbrowser/min#630]
I am absolutely in love with this browser - yet, I can't use it. The unnecessary strip of empty pixels in the upper section makes the user interface extremely hard to use. Using any of the user elements in the upper part (which is all of them) now requires quite a precise mouse work, which really breaks ones workflow.

If I could truly pin the windows to the top of screen so that I could just move to mouse to the upper edge and still be able to click that, that would be all I need.

In other words - please, get rid of the strip of empty pixels above the buttons, tabs and other user elements.
Thank you and keep up the great work.

**Comment by Anavnaj** [comment]
I'm adding an image just to make myself perfectly clear about the issue.

![pixelstrip](https://user-images.githubusercontent.com/43697644/46244252-79e3e500-c3dc-11e8-83b7-404329f2aa06.png)

**Comment by PalmerAL** [comment]
> A) Make the space to the right of the tabs (or, rather, to the left of the "+" button, or perhaps even of the "minimize window" button, as seen in the linked image) really small - I don't see a reason why it should be wider that 3-4 times the width of the cursor. This will not limit the space to see the open tabs significantly...

You mean something like this (where the red is the draggable area)?

<img width="777" alt="screen shot 2018-10-02" src="https://user-images.githubusercontent.com/10314059/46330020-6dc37780-c5d6-11e8-80a9-dc3c445eabd6.png">

That actually works better than I would have expected, although it does look a little bit odd to have a space between the two buttons. 

> B) Have the upper strip of pixels appear only when the browser window is not maximized. When the browser window is maximized, there's no real need for dragging it, right?* I mean - doing so "unmaximizes" it anyway.

This seems pretty reasonable to me, although the snapping thing you mentioned may be an issue. I also wonder if it would be confusing whether the window is actually maximized (in which case you would need to click the button first), or whether it was almost-but-not-quite maximized (in which case you shouldn't click the button and need to drag from the top instead).

Another thing I wonder about is whether not having the tabs at the top is actually the issue, or whether it's just the position of the draggable area and the height of the tabs. Currently, the draggable area extends a little bit lower than you would guess:

<img width="805" alt="screen shot 2018-10-02 2" src="https://user-images.githubusercontent.com/10314059/46330119-e7f3fc00-c5d6-11e8-9a41-31e00391c8c7.png">

(In particular, the bottom of the drag region extends a little bit past the top of the window buttons).
If we made the whole navbar a little bit taller, we could increase the click area of the tabs, and make it line up with where they look like they end, which might make it easier to position the mouse.

**Comment by Anavnaj** [comment]
Actually, my problem with not having the tabs at the top comes from being used to just shoot my cursor in the general direction of the tab I want to switch to without having to aim precisely on the vertical line. With the space above the tabs, this is not possible, as I am forced to carefully slide the cursor a bit below the top (but not too much, considering the tabs aren't that wide themselves - which is otherwise a great thing!). This is especially bothersome with a touchpad.

**Ad A)**
I would strongly advocate for putting the draggable area actually to the right of the "tasks" button. I think it makes sense to have it between the two specific "Min" buttons and the universal Windows "minimize/maximize/close" buttons, and it also lowers the risk of accidentally minimizing the browser when the user actually wanted to click the "tasks" button. Last but not least, it might also look better (considering those two types of buttons have noticeably different design - which I think is a good thing, but also a reason why dividing them by the clean space of the draggable area might make sense aesthetically).

**Ad B)**
I see your point, or rather both your points. As for the potential "snapping" issue - I'm fairly sure that vast majority of people mainly use the "Win Button + arrow key" keyboard shortcut, especially when snapping maximized Windows, so having to "unmaximze" the browser first in order to drag-snap it shouldn't affect many users, and even in their case, I'd argue it's not that much of a bother. In any case, this problem would be completely solved by combining the option B) with option A).

As for your other point - I think that discerning between maximized and "almost-but-not-quite maximized" windows might potentially be a problem only in case the user has just one tab opened (in case of two or more tabs, the color difference makes the upper strip immediately visible), and even then, one can still tell by how the "maximize/unmaximize" button looks...


I also realized there's an **option C)** [that is also compatible with at least the option A)] utilized by some of the mainstream browsers (Vivaldi, Opera), and that is making the strip above the tabs optional when the browser is maximized (i.e. giving the user the option to turn it on/off in the settings). Now that I think of it, making option A also optional might make a lot of sense, too (if possible and viable).

So, I guess that _my_ perfect solution would be combining all three options and making options A) and C) optional in the settings. I don't see any real downside to this, but I also have no idea how difficult would be to implement it (although I was impressed how quickly did you "demo" the two previously discussed options, so maybe not that difficult?). In case you're not too keen on cluttering the settings with to many options (and also possibly complicating future development with too much of a "user interface" diversity), I guess combining option A) and B) should also satisfy the needs of all kinds of users...

Once again - thanks for the time your putting into this!

**Comment by PalmerAL** [comment]
> Actually, my problem with not having the tabs at the top comes from being used to just shoot my cursor in the general direction of the tab I want to switch to without having to aim precisely on the vertical line. With the space above the tabs, this is not possible, as I am forced to carefully slide the cursor a bit below the top (but not too much, considering the tabs aren't that wide themselves - which is otherwise a great thing!). This is especially bothersome with a touchpad.

OK, makes sense! I checked what other browsers do, and it seems like they also take this into account. Firefox always has the tabs at the top of the window; Chrome has a dragging region similar to what we do, but hides it when the window is maximized. (Interestingly, both Chrome and Firefox seem to have tabs that are even shorter than ours - do they seem shorter because they look different visually, or because they're difficult to click)?

Having the drag area between the task and minimize buttons definitely is better aesthetically, but there's a couple of issues with it:

* When the task overlay is open, there's a weird gap between the buttons that doesn't obviously look like a draggable area.
* There's less room for the searchbar, which makes it harder to edit a long URL, particularly when the window is narrow.
  * We can solve this by hiding the drag area when the searchbar is open, but this makes the buttons jump around in kind of an odd way, particularly if you try to open the overlay while the searchbar is still open (as seen in the attached gif).
  * We can also hide the buttons entirely when the searchbar is open, which looks better. However, given that the new tab and task buttons are one of the main ways of interacting with the browser, I wonder if it would be confusing to have them disappear unexpectedly.

![drag-area-right](https://user-images.githubusercontent.com/10314059/46450495-23b5cf80-c757-11e8-8744-11bb7e0f37d6.gif)

Of these, option B does sound like the best option; hopefully people will be able to realize when the window is maximized. I'm not sure I see a reason to have a preference here - I'd rather come up with one design that works well rather than have multiple ones with different flaws (both because that's easier to implement, and also because it makes the browser easier to use).

Having the navbar height change depending on whether the window is maximized or not is going to be kind of a pain, so I probably won't get to that for a bit (unless someone else wants to work on it first). For 1.8, I think I'm going to implement the solution of shifting the draggable area up slightly when the window is maximized. This should help a bit by reducing the amount you have to move the mouse down, and increasing the height of the tab area, and I can't think of any reason not to do it. Hopefully, we should be able to implement the full solution in the version or two after that.

**Comment by PalmerAL** [comment]
Actually, it would probably be easier to just go ahead and fix this completely for 1.9, and not do anything for 1.8.

**Comment by ChildishGiant** [comment]
I'd also love for this minor gripe to be fixed 😄

**Comment by PalmerAL** [comment]
@lordhellraiser I don't think there's any way to do that, at least with Electron. Any areas that we mark as draggable are treated as [non-client frames](https://electronjs.org/docs/api/frameless-window#context-menu), so there's no way to detect mouse events within them. We could possibly implement our own window dragging in JS, but that's probably going to cause tons of bugs with things like aero snap.

**Comment by lordhellraiser** [comment]
It's a shame that Windows doesn't support this. I use Authotkey to have that functionality. Another option is to remove those pixels while maximized. What you think?

**Comment by PalmerAL** [comment]
@lordhellraiser Yeah, I agree that's probably the best option.

**Comment by PalmerAL** [comment]
I've modified the title bar so that the drag area only appears when the window isn't maximized in bd5ec6850b47d862c760f8304bd1e144ee6de90a; would you be willing to test this and confirm that everything works as expected?

**Comment by ChildishGiant** [comment]
I'm too lazy to build it but I'll happily test a build

**Comment by PalmerAL** [comment]
@ChildishGiant Here you go: https://send.firefox.com/download/8901bac596/#rzBV5eIgSc4OEU5RVZhJzA

**Comment by ChildishGiant** [comment]
Much better!

**Comment by moigagoo** [comment]
Tried id too. So much better, thanks @PalmerAL!

**Comment by LukaFX** [comment]
honestly I don't like this change. I like to be able to drag the windows down when its fullscreen even if you cant drag from the top.

**Comment by PalmerAL** [comment]
@ChildishGiant @moigagoo I'm glad everything's working!

@LukaFX Another option I considered was adding a draggable space on the left side of the window:

<img width="800" alt="dragspaceleft" src="https://user-images.githubusercontent.com/10314059/50495534-4fabbe80-09ef-11e9-82f7-93b836eacb4b.png">

I'm not sure if this is actually a good idea or not (in particular, it means you can't reach the menu button by moving your mouse to the upper-left corner), but it would still allow you to drag the window while solving the first issue. Thoughts?

**Comment by LukaFX** [comment]
@PalmerAL that would be better for me because I have two monitors, mabye make it a setting or something?

**Comment by ChildishGiant** [comment]
I have two monitors but use most applications maximised so don't drag much.

**Comment by PalmerAL** [comment]
@LukaFX I don't think a setting is the right answer for this; assuming there aren't any major objections to this, it's probably better to just do it by default.
